Streaming

llfw opened the pull request:
note: this depends on #5.

add streaming (push) support.  this is using the HTTP (non-WebSocket) streaming interface for now.

the bulk of the implementation is in `api.streaming`, which provides the basic streaming client `HTTPStreamClient`.  however, because streams don't map directly onto Timelines (e.g., the `user` stream contains both `HomeTimeline` and `NotificationTimeline`), we need a way to (de)multiplex the stream and handle multiple subscribers.  this is handled by `StreamMultiplexer`, which manages a list of subscribers for each stream and distributes events to them as needed.  streams are created on demand, and when the last subscriber exits, the stream is closed.

the public API is provided by `StreamSubscription`, which is a handle to a stream.  `api.timeline` uses this in the `streaming()` method to fetch events from the stream and dispatch them to the timeline's queue.

add a new `--streaming` cli/config option to enable streaming; disabled by default.

Commit:
api: add streaming support

Add an InstanceStreamer class which manages streaming for one instance.
Consumers can subscribe to a stream by calling subscribe(), which
returns a StreamSubscription.  Events received on the stream will be
posted to the StreamSubscription's queue member.

Streaming subscribers are refcounted, so any number of users subscribing
to the same stream will only create one HTTP streaming connection.

Commit:
api.timeline: support streaming in StatusTimeline

Add support for streaming in StatusTimeline using the new api.streaming
interface.  For now, only FederatedTimeline has streaming enabled.

Commit:
app: add streaming support

Add a new flag, --streaming, which enables streaming timelines.  Teach
TimelineTab to enable streaming if supported.

Commit:
api.timeline: handle duplicate events

When calling update(), ignore any events we received via streaming to
avoid posting duplicate events.  This ensure we'll still fetch any
events that streaming missed.

Commit:
app: if streaming is enabled, enable timeline_refresh

Streaming is not reliable, so we need to enable timeline_refresh to
catch any events we missed.  Use 120 seconds as the default, which
should be reasonable quick without creating excessive network traffic.

Commit:
timeline: enable streaming for HomeTimeline and LocalTimeline

Commit:
timeline: lift can_stream into Timeline

There's no need for this to be a property since it can't change; make it
a constructor argument of Timeline instead.

Commit:
api.timeline: enable streaming for NotificationTimeline

Commit:
Timeline: hold _lock in close() and streaming()

This avoids a possible race if we're closed while streaming is being
enabled.

Commit:
api.streaming: handle errors from the stream endpoint

If we get an error from the stream while streaming, wait for
ERROR_BACKOFF (30) seconds, then reconnect.

ihabunek commented on tooi/api/streaming.py:
```diff
@@ -0,0 +1,225 @@
+# Streaming (push) support for timelines.  Streaming is handled by the InstanceStreamer class, which
+# should only be created once per instance.  InstanceStreamer handles connecting to the streaming
+# endpoint, receiving events, and distributing them to the various subscribers.  To subscribe to
+# streaming, use subscribe() to create a StreamSubscription instance.
```
It's customary to put module docs at the very top of the file in a triple-quoted string.

llfw replied:
fixed in 4490988

ihabunek commented on tooi/api/streaming.py:
```diff
@@ -0,0 +1,225 @@
+# streaming, use subscribe() to create a StreamSubscription instance.
+
+
+class StreamEvent(object):
```
Nitpick: no need to inherit `object` since py3. You can leave out the parens too.

llfw replied:
fixed in 4490988

ihabunek commented on tooi/api/streaming.py:
```diff
@@ -0,0 +1,241 @@
+    async def _stream(self):
+        logger.info("HTTPStreamClient: connecting to stream={stream}")
+
+        async with self.ctx.auth.client.stream("GET", self.url, timeout=self.TIMEOUT) as stream:
```
You should `stream.raise_for_status()` here so that we don't attempt to process failed requests.

Additionally, the try/catch above should catch the resulting exception (probably catch HTTPError instead of RequestError).

In the long run, it would be good to have our variant of `stream` in the `tooi.api` module like we have `request`.

llfw replied:
fixed in ba404607bce4e9656017c7471864fa47a9013c45

ihabunek commented on tooi/api/streaming.py:
```diff
@@ -0,0 +1,241 @@
+                await asyncio.sleep(self.ERROR_BACKOFF)
+
+    async def _stream(self):
+        logger.info("HTTPStreamClient: connecting to stream={stream}")
```
```suggestion
        logger.info(f"HTTPStreamClient: connecting to stream={self.stream_name}")
```

llfw replied:
fixed in 6025178a3b3ce26093985013d8b9f3d6bc86e824

ihabunek commented on tooi/api/streaming.py:
```diff
@@ -0,0 +1,241 @@
+    #         or: StreamInstance.lock if self.lock not held
+    #        NOT: StreamInstance.lock -> self.lock
+
+    class StreamInstance(object):
```
Can this class not be nested? I don't see any advantages.

llfw replied:
fixed in 4490988977125d6815145d9f255e2168a2287990

ihabunek reviewed (commented):
Some preliminary comment, didn't have time to finish review.

Commit:
api.streaming: style cleanups

- use a doc comment for file header
- make StreamInstance not nested
- don't explicitly inherit from object

Commit:
HTTPStreamClient: fix log message

Commit:
HTTPStreamClient: improve error handling

- call raise_for_status() on the stream so we don't try to stream errors
- back off any time the stream is closed, to avoid reconnecting in a
  loop.  the stream is never supposed to end, so even a clean close
  probably indicates the server is restarting or something like that.
- catch HTTPError instead of RequestError

Commit:
api.streaming: implement the protocol correctly

The Mastodon API docs are not very clear about this, but according to
the MDN documentation[0] each event can be multiple lines in length and
will be terminated by a blank line.  Fix _handle_line() to implement the
protocol like this, which fixes occasional issues with event payloads
that had embedded newlines.

[0] https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events

ghost commented:
rebased on main (678edd0e320027dfd3bcf700ca4cba685f12585f)

Commit:
api.timeline: handle duplicate statuses in stream()

If the server pushes an event to the stream, and we call _update()
before we handle it, and the event is also returned by the update, then
it will be added to the timeline twice.  Fix this by tracking events
received by both updates and streaming, and eliding duplicates in either
case.

In theory this could allow self._seen_events to grow forever if _update
is never called, but we forcibly enable timeline_refresh if streaming is
enabled, so that should never happen.

Commit:
InstanceInfo: add streaming_url()

This returns the streaming URL for this instance.

While here, rename InstanceStatusConfguration to
InstanceStatusConfiguration.

Commit:
auth: support aiohttp

For now we don't cache the aiohttp ClientSession, because it's a bit
awkward to create one: we have to do it inside an async function so we
can pass Textual's async loop context, otherwise aiohttp creates its own
and doesn't work.

Commit:
HTTPStreamClient: use aiohttp instead of httpx

Commit:
HTTPStreamClient: handle UTF-8 errors

Commit:
auth: cache the aiohttp.ClientSession

Commit:
api.streaming: support WebSocket protocol

Add WSStreamClient which supports the WebSocket protocol.  Choose the
appropriate client to start based on the server's reported streaming
URL.

ihabunek commented on tooi/api/streaming.py:
```diff
@@ -0,0 +1,392 @@
+
+        # Since the base URL of the stream might be different from the instance's normal base URL,
+        # we have to create our own client.
+        client = aiohttp.ClientSession()
```
Why not use the context manager here as they recommend?

```py
async with aiohttp.ClientSession() as session:
```

llfw replied:
because i forgot :-)  fixed in 5b2b6ca

ihabunek commented on tooi/api/streaming.py:
```diff
@@ -0,0 +1,392 @@
+
+        logger.info(f"WSStreamClient: url={url} for stream={self.stream_name}")
+
+        async with client.ws_connect(url, params=params, timeout=timeout) as resp:
```
`ws_connect` expects a float value for timeout, the timeout struct is meant to be passed to the `ClientSession` constructor.

llfw replied:
fixed in 9d8126b

ihabunek commented on tooi/api/streaming.py:
```diff
@@ -0,0 +1,392 @@
+            del self.streams[stream]
+
+    def run(self):
+        self._task = run_async_task(self._run())
```
I don't think this is used, since `StreamMultiplexer` doesn't have a `_run` method.

llfw replied:
fixed in a9e6d8f

ihabunek committed:
Improve types

The asserts stop the type checker from complaining if a variable is
nullable but we know it won't be null at that point.

ihabunek reviewed (commented):
A few more comments.

Commit:
api.streaming: fix public timelines in WSHttpClient

The WS endpoint uses ':' for these (public:local) while the HTTP
endpoint uses '/' (public/local).

Commit:
api.streaming: use context manager in WSHttpClient

Commit:
api.streaming: pass aiohttp's timeout correctly

Commit:
api.streaming: remove unused method

StreamMultiplex.run() was left over from older code.

ihabunek committed:
Avoid refreshing status details if it has not changed

This fixes an issue that status details was redrawn when new events were
added via streaming. The cause was that ListView was sending the
ListView.Highlighted event even thought the focused item has not
changed.

ihabunek committed:
Handle unlikely scenario where instance v1 is null
